File: server/lib/peak_detection.py
import numpy as np
from numpy import ndarray
from scipy.signal import find_peaks


# 64電極すべての下ピークを取得
def detect_peak_neg(
    MEA_data: ndarray,
    distance=3000,
    threshold=3,
    min_amp=10,
    width=None,
    prominence=None,
) -> ndarray:
    peak_index = np.array([None for _ in range(len(MEA_data))])
    for i in range(1, len(MEA_data)):
        # ピーク抽出の閾値を設定
        height = np.std(MEA_data[i]) * threshold
        # 閾値が最低閾値を下回っていた場合は最低閾値の値を閾値の値に設定する
        if height < min_amp:
            height = min_amp
        detect_peak_index = find_peaks(
            -MEA_data[i],
            height=height,
            distance=distance,
            width=width,
            prominence=prominence,
        )

        peak_index[i] = detect_peak_index[0]
        peak_index[i] = np.sort(peak_index[i])
    peak_index[0] = np.array([])

    return peak_index


# 64電極すべての上ピークを取得
def detect_peak_pos(
    MEA_data: ndarray,
    distance=3000,
    threshold=3,
    min_amp=10,
    width=None,
    prominence=None,
) -> ndarray:
    peak_index = np.array([None for _ in range(len(MEA_data))])
    for i in range(1, len(MEA_data)):
        # ピーク抽出の閾値を設定
        height = np.std(MEA_data[i]) * threshold
        # 閾値が最低閾値を下回っていた場合は最低閾値の値を閾値の値に設定する
        if height < min_amp:
            height = min_amp
        detect_peak_index = find_peaks(
            MEA_data[i],
            height=height,
            distance=distance,
            width=width,
            prominence=prominence,
        )

        peak_index[i] = detect_peak_index[0]
        peak_index[i] = np.sort(peak_index[i])
    peak_index[0] = np.array([])

    return peak_index

File: go-backend/lib/peakDetection.go
package lib

import (
	"sort"
	"sync"

	"gonum.org/v1/gonum/stat"
)

type PeakDetection struct {
	MeaData [][]float32
}

func NewPeakDetection(meaData [][]float32) *PeakDetection {
	return &PeakDetection{MeaData: meaData}
}

func (pd *PeakDetection) DetectPeakNeg(distance, threshold, minAmp int) [][]int {
	peakIndex := make([][]int, len(pd.MeaData))

	var wg sync.WaitGroup
	ch := make(chan struct{}, 8)
	var mu sync.Mutex

	for i := 1; i < len(pd.MeaData); i++ {
		wg.Add(1)
		ch <- struct{}{}
		go func(i int) {
			defer wg.Done()
			defer func() { <-ch }()
			data := make([]float64, len(pd.MeaData[i]))
			for i, d := range pd.MeaData[i] {
				data[i] = float64(d)
			}
			height := stat.StdDev(data, nil) * float64(threshold)
			if height < float64(minAmp) {
				height = float64(minAmp)
			}

			peaks := findPeaksNeg(data, height, distance)
			sort.Ints(peaks)
			mu.Lock()
			peakIndex[i] = peaks
			mu.Unlock()
		}(i)

	}
	wg.Wait()
	return peakIndex
}

func (pd *PeakDetection) DetectPeakPos(distance, threshold, minAmp int) [][]int {
	peakIndex := make([][]int, len(pd.MeaData))

	var wg sync.WaitGroup
	ch := make(chan struct{}, 8)
	var mu sync.Mutex

	for i := 1; i < len(pd.MeaData); i++ {
		wg.Add(1)
		ch <- struct{}{}
		go func(i int) {
			defer wg.Done()
			defer func() { <-ch }()
			data := make([]float64, len(pd.MeaData[i]))
			for i, d := range pd.MeaData[i] {
				data[i] = float64(d)
			}
			height := stat.StdDev(data, nil) * float64(threshold)
			if height < float64(minAmp) {
				height = float64(minAmp)
			}

			peaks := findPeaksPos(data, height, distance)
			sort.Ints(peaks)
			mu.Lock()
			peakIndex[i] = peaks
			mu.Unlock()
		}(i)

	}
	wg.Wait()
	return peakIndex
}

func findPeaksNeg(data []float64, height float64, distance int) []int {
	negData := make([]float64, len(data))
	for i := range data {
		negData[i] = -data[i]
	}

	peaks := []int{}
	for i := 1; i < len(negData)-1; i++ {
		if negData[i] > negData[i-1] && negData[i] > negData[i+1] && negData[i] >= height {
			if len(peaks) == 0 || i-peaks[len(peaks)-1] >= distance {
				peaks = append(peaks, i)
			}
		}
	}

	// 幅とプロミネンスに基づく追加フィルタリングをここに追加できます

	return peaks
}

func findPeaksPos(data []float64, height float64, distance int) []int {
	peaks := []int{}
	for i := 1; i < len(data)-1; i++ {
		if data[i] > data[i-1] && data[i] > data[i+1] && data[i] >= height {
			if len(peaks) == 0 || i-peaks[len(peaks)-1] >= distance {
				peaks = append(peaks, i)
			}
		}
	}

	// 幅とプロミネンスに基づく追加フィルタリングをここに追加できます

	return peaks
}
